### PR TITLE
lift the `=dup` hook

### PR DESCRIPTION
**Pull request by ringabout**

### CODE REVIEW

File: compiler/liftdestructors.nim [hunk 1]
```diff
@@ -861,16 +890,16 @@ proc ownedClosureOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
   #discard addDestructorCall(c, elemType, newNodeI(nkStmtList, c.info), genDeref(xx))
   actions.add callCodegenProc(c.g, "nimDestroyAndDispose", c.info, xx)
   case c.kind
-  of attachedSink, attachedAsgn:
+  of attachedSink, attachedAsgn, attachedDup:
+    if c.kind == attachedDup:
+      body.add genBuiltin(c, mWasMoved, "`=wasMoved`", x)
```
Comment (Araq) [inline compiler/liftdestructors.nim]: It is wrong to treat dup like assignment. And a dup never requires a `wasMoved`, that's the point of dup.

Comment (ringabout) [reply compiler/liftdestructors.nim]: Closures aside, How should I implement `=dup` for non-ref types?

Comment (ringabout) [reply compiler/liftdestructors.nim]: I have removed `wasMoved`. Can I treat them like assignments for non-ref types?

Comment (Araq) [reply compiler/liftdestructors.nim]: No. The way a lifted dup looks like for an object is:

```

proc dup(x: Obj): Obj = Obj(a: dup(x.a), b: dup(x.b))

```

quite different from the other lifted procs. And a challenge for case objects, as usual.

Comment (ringabout) [reply compiler/liftdestructors.nim]: Yeah, it is for objects
```nim
type
  Chuck = object
  Task = object
    id: Chuck
    name: string

proc `=dup`(x: Chuck): Chuck =
  echo "Chuck"
```


```c
N_LIB_PRIVATE N_NIMCALL(tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ, eqdup___test54_u92)(tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ src_p0)
{
	tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ result;
	tyObject_Chuck__kwZ1UmFHwLaBemQzcUHiDA T1_;
	NIM_BOOL *nimErr_;
	{
		nimErr_ = nimErrorFlag();
		nimZeroMem((void *)(&result), sizeof(tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ));
		nimZeroMem((void *)(&T1_), sizeof(tyObject_Chuck__kwZ1UmFHwLaBemQzcUHiDA));
		T1_ = eqdup___test54_u5(src_p0.id);
		if (NIM_UNLIKELY(*nimErr_))
			goto BeforeRet_;
		result.id = T1_;
		result.name = eqdup___stdZassertions_u36(src_p0.name);
	}
BeforeRet_:;
	return result;
}
```
The lifted `=dup` hook for `Task` calls its object field's `dup` hook. It uses `nimAsgnStr` for its string field.

Comment (ringabout) [reply compiler/liftdestructors.nim]: > The way a lifted dup looks like for an object is:

This is how it is lifted for objects

```nim
proc dup(x: Obj): Obj =
  result.a = `=dup`(x.a)
  result.b = `=dup`(x.b)
```
I assume it's good enough, isn't it?

Comment (ringabout) [reply compiler/liftdestructors.nim]: Is this right for object variants?

```nim
proc `=dup`(x: Variant): Variant =
  result.kind = x.kind
  case x.kind
  of a:
    result.a1 = `=dup`(x.a1)
  of b:
    result.b1 = `=dup`(x.b1)
  of c:
    result.c1 = `=dup`(x.c1)
```

Comment (Araq) [reply compiler/liftdestructors.nim]: > I assume it's good enough, isn't it?

Only barely. But you can improve it later.

> Is this right for object variants?

Why not `result = Variant(kind: x.kind); case ...`

Comment (ringabout) [reply compiler/liftdestructors.nim]: > Why not result = Variant(kind: x.kind); case ...

I don't need to make any changes to generate the former code. The same applies to lift it for objects. Isn't it a pre-existing issue (It's what `=copy` does)? I think I can improve it later.

**Comment by Araq** [comment]
Any benchmark improvements?

File: compiler/liftdestructors.nim [hunk 2]
```diff
@@ -1033,9 +1105,11 @@ proc produceSym(g: ModuleGraph; c: PContext; typ: PType; kind: TTypeAttachedOp;
   var a = TLiftCtx(info: info, g: g, kind: kind, c: c, asgnForType: typ, idgen: idgen,
                    fn: result)
 
-  let dest = result.typ.n[1].sym
-  let d = newDeref(newSymNode(dest))
-  let src = if kind in {attachedDestructor, attachedWasMoved, attachedDup}: newNodeIT(nkSym, info, getSysType(g, info, tyPointer))
+  let dest = if kind == attachedDup: result.ast[resultPos].sym else: result.typ.n[1].sym
```
Comment (Araq) [inline compiler/liftdestructors.nim:1108]: Why is

```nim

let dest = result.typ.n[1].sym

```

not good enough for `attachedDup`?

File: lib/system/arc.nim [hunk 3]
```diff
@@ -200,10 +200,6 @@ proc nimDecRefIsLast(p: pointer): bool {.compilerRtl, inl.} =
       when traceCollector:
         cprintf("[DECREF] %p\n", cell)
 
-proc nimDupRef(dest: ptr pointer, src: pointer) {.compilerRtl, inl.} =
-  dest[] = src
-  if src != nil: nimIncRef src
-
```
Comment (Araq) [inline lib/system/arc.nim:206]: How can `=dup` work for `ref T` without something like `nimDupRef`?

File: compiler/liftdestructors.nim (hunk 1 above)
Comment (Araq) [reply compiler/liftdestructors.nim]: Yes, you can.

**Comment by Araq** [comment]
Awesome, clean solution! But please answer my questions.

File: lib/system/arc.nim (hunk 3 above)
Comment (ringabout) [reply lib/system/arc.nim:206]: I use the direct code gen combination in `atomicRefOp`

```nim
    if isCyclic:
      body.add newAsgnStmt(x, y)
      body.add genIf(c, y, callCodegenProc(c.g,
          "nimIncRefCyclic", c.info, y, getCycleParam(c)))
    else:
      body.add newAsgnStmt(x, y)
      body.add genIf(c, y, callCodegenProc(c.g,
          "nimIncRef", c.info, y))
```

File: compiler/liftdestructors.nim (hunk 2 above)
Comment (ringabout) [reply compiler/liftdestructors.nim:1108]: Here is the signature of a dup hook

```nim
proc `=dup`[T](src: T): T =
```

`result.typ.n[1].sym` is `src`.

Comment (Araq) [reply compiler/liftdestructors.nim:1108]: I misread, thought it's about the involved types (where it doesn't matter).

**Pull request merged**